src/server/game/Globals/LoadDataQueries.cpp

```diff
@@ -0,0 +1,1230 @@
+#include "LoadDataQueries.h"
+#include "Errors.h"
+#include <vector>
+
+static std::vector<LoadDataQuery*> asyncQueries;
+
+LoadDataQuery::LoadDataQuery(std::string const& query)
+    : m_query(query)
+{}
+
+QueryResult LoadDataQuery::GetOrQueryResults()
+{
+    WaitForAsyncQuery();
+    if (m_asyncResult.has_value())
```

Comment:
can race condition occur here if GetOrQueryResults called the same time callback executed?

Reply:
> can race condition occur here if GetOrQueryResults called the same time callback executed?

Currently, the callback is only executed from inside `WaitForAsyncQuery` (that's the only place that calls `InvokeIfReady`) so as long as any callers are single-threaded (and they are, assuming we start these queries before any tables load), that should not be possible.